### PR TITLE
Upgrade dependencies 2024-02-19 (#5970)

### LINKED ISSUE #5970: Upgrade dependencies 2024-02-19
- [x] Update [PyCharm image](https://github.com/DataBiosphere/azul-docker-pycharm)
  - [x] Bump [base image](https://hub.docker.com/_/debian/tags?name=bullseye) tag (only same Debian release), if possible
  - [x] Bump upstream version, if possible
  - [x] Bump internal version
  - [ ] ~Build and test new image locally with Azul's `make format`~
  - [x] Remove unused dependencies with high or critical CVEs
  - [x] Push commit to GitHub (directly to master branch, no PR needed)
  - [x] GH Action workflow succeeded
  - [x] Image is available on [DockerHub](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm) 
- [x] Update [Elasticsearch image](https://github.com/DataBiosphere/azul-docker-elasticsearch)
  - [x] Bump [base image](https://hub.docker.com/_/elasticsearch/tags) tag (only minor and patch versions), if possible
  - [x] Bump internal version 
  - [ ] ~Build and test new image locally with Azul's `make test`~
  - [x] Remove unused dependencies with high or critical CVEs
  - [x] Push commit to GitHub (directly to master branch, no PR needed)
  - [x] GH Action workflow succeeded
  - [x] Image is available on [DockerHub](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch) 
- [x] Create Azul PR, connected to this issue, with … 
  - [x] … changes to `requirements*.txt` from open Dependabot PRs, one commit per PR
  - [x] … update to [Python](https://hub.docker.com/_/python/tags) (only patch versions) <sub>or no update available</sub>
  - [x] … update to [Terraform](https://hub.docker.com/r/hashicorp/terraform/tags) (only patch versions) <sub>or no update available</sub>
  - [x] … new [PyCharm image](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm)
  - [x] … new [Elasticsearch image](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch)
  - [x] … update to [Docker images](https://hub.docker.com/_/docker/tags) (only minor and patch versions) <sub>or no update available</sub>
  - [x] … update to [GitLab](https://hub.docker.com/r/gitlab/gitlab-ce/tags) & [GitLab runner images](https://hub.docker.com/r/gitlab/gitlab-runner/tags) <sub>or no update available</sub>
  - [x] … update to [ClamAV image](https://hub.docker.com/r/clamav/clamav/tags) <sub>or no update available</sub>
  - [x] … update to [GitLab AMI](https://github.com/DataBiosphere/azul/blob/develop/OPERATOR.rst#updating-the-ami-for-gitlab-instances) <sub>or no update available</sub>
- [ ] ~Delete obsolete image tags from DockerHub (but consider that `prod` may not use the latest image)~ … (This is now done by System administrator during `prod` promotion)
  - [ ] … ~for [PyCharm image](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm)~
  - [ ] … ~for [Elasticsearch image](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch)~
- [ ] Created tickets for any deferred updates to …
  - [x] … to next major or minor Python version <sub>or such ticket already exists</sub> https://github.com/DataBiosphere/azul/issues/5736
  - [x] … to next major Docker version <sub>or such ticket already exists</sub> https://github.com/DataBiosphere/azul/issues/5865
  - [x] … to next major or minor Terraform version <sub>or such ticket already exists</sub> https://github.com/DataBiosphere/azul/issues/5893

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issue: #5970

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `upgrades/yyyy-mm-dd`
- [x] On ZenHub, PR is connected to the upgrade issue it resolves
- [x] PR title matches `Upgrade dependencies yyyy-mm-dd`
- [x] PR title references the connected issue

### Author (upgrading deployments)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] Ran `make image_manifests.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `image_manifests.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `no demo`
- [x] A comment to this PR details the completed security design review <sub>or this PR is a promotion or a backport</sub>
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvilprod.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvilprod.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] Assigned system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>

### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvilprod.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvilprod.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label
- [x] Pushed PR branch to GitLab `dev`
- [x] Pushed PR branch to GitLab `anvildev`
- [x] Pushed PR branch to GitLab `anvilprod`
- [x] Build passes in `sandbox` deployment
- [x] Build passes in `anvilbox` deployment
- [x] Build passes in `hammerbox` deployment
- [x] Reviewed build logs for anomalies in `sandbox` deployment
- [x] Reviewed build logs for anomalies in `anvilbox` deployment
- [x] Reviewed build logs for anomalies in `hammerbox` deployment
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but excluded any `p` tags</sub>
- [x] Moved connected issue to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Pushed merge commit to GitLab `anvilprod`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Build passes on GitLab `anvilprod`
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvilprod.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

### Operator

- [x] Ran `script/export_inspector_findings.py` against `anvilprod`, imported results to [Google Sheet](https://docs.google.com/spreadsheets/d/1RWF7g5wRKWPGovLw4jpJGX_XMi8aWLXLOvvE5rxqgH8) and posted screenshot of relevant<sup>1</sup> findings as a comment on the connected issue.
- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `reindex:partial` and `reindex:prod` labels to the next promotion PR <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `reindex:partial` and `reindex:prod` labels from the description of this PR to that of the next promotion PR <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to only the system administrator

<sup>1</sup>A relevant finding is a high or critical vulnerability in an image
that is used within the security boundary. Images not used within the boundary
are tracked in `azul.docker_images` under a key starting with `_`.

### System administrator

- [x] No currently reported vulnerability requires immediate attention
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by dsotirho-ucsc** [comment]
[5996_IT_2024-02-28.txt](https://github.com/DataBiosphere/azul/files/14440863/5996_IT_2024-02-28.txt)

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please update the PR checklist to the one from the recently updated template.

**Review by hannes-ucsc — COMMENTED** [review]
Conflicts.

**Comment by dsotirho-ucsc** [comment]
[5996_IT_2024-03-01.txt](https://github.com/DataBiosphere/azul/files/14465502/5996_IT_2024-03-01.txt)

File: UPGRADING.rst [hunk 1]
```diff
@@ -20,6 +20,26 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
...
+Makefile target for the lower deployments after pushing the squashed and rebased
+PR branch to GitHub. In a stable deployment (``prod``), perform the first phase
+before pushing the merge commit to the GitLab instance in that deployment. In
+lower and stable deployments, perform the second phase using the ``apply``
+Makefile target after the merge commit was successfully built on the GitLab
+instance in that deployment.
+
+Deploy the ``gitlab`` component of any main deployment just after pushing the PR
+branch to GitHub. Run ``make -C terraform/gitlab/runner`` just before pushing
+the merge commit to the GitLab instance in that deployment.
```
Comment (hannes-ucsc) [inline UPGRADING.rst:40]: The checklist changes make this unnecessary.

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by hannes-ucsc** [comment]
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

**Comment by hannes-ucsc** [comment]
https://github.com/DataBiosphere/azul/issues/4890 has occurred on `anvildev.gitlab`. Please apply the workaround. Remember to check whether GitLab is functional after deploying it.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**